packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.kt

```diff
@@ -80,6 +90,72 @@ class FileSystemModule : Module() {
       return@Coroutine destination.toURI()
     }
 
+    var pendingPickerPromise: Promise? = null
+
+    OnActivityResult { _, (requestCode, resultCode, data) ->
+      if (pendingPickerPromise == null) return@OnActivityResult
+      if (requestCode != PICKER_REQUEST_DIRECTORY && requestCode != PICKER_REQUEST_FILE) return@OnActivityResult
+      if (resultCode != Activity.RESULT_OK || data == null) {
+        pendingPickerPromise?.reject(PickerCancelledException())
+        pendingPickerPromise = null
+        return@OnActivityResult
+      }
+      val uri = data.data
+      val takeFlags = (
+        data.flags
+          and (Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_WRITE_URI_PERMISSION)
+        )
+      uri?.let {
+        appContext.throwingActivity.contentResolver.takePersistableUriPermission(it, takeFlags)
+      }
+      when (requestCode) {
+        PICKER_REQUEST_DIRECTORY -> {
+          pendingPickerPromise?.resolve(
+            FileSystemDirectory(uri ?: Uri.EMPTY)
+          )
+        }
+
+        PICKER_REQUEST_FILE -> {
+          pendingPickerPromise?.resolve(
+            FileSystemFile(uri ?: Uri.EMPTY)
+          )
+        }
+      }
+      pendingPickerPromise = null
+    }
+
+    AsyncFunction("pickDirectoryAsync") { initialUri: Uri?, promise: Promise ->
+      if (pendingPickerPromise != null) {
+        throw FileSystemPendingPickerException()
+      }
+      val intent = Intent(Intent.ACTION_OPEN_DOCUMENT_TREE)
+//        intent.addCategory(Intent.CATEGORY_OPENABLE)
+      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+        initialUri
+          .let { intent.putExtra(DocumentsContract.EXTRA_INITIAL_URI, it) }
+      }
+
+      pendingPickerPromise = promise
+      appContext.throwingActivity.startActivityForResult(intent, PICKER_REQUEST_DIRECTORY)
```

Comment:
Can we use contracts instead?
https://github.com/expo/expo/blob/da7cdc0d25890225d0c4639f7e0e4cc65e46ede2/packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerModule.kt#L121-L123

Reply:
oh dude, I didn't recall we had those and I was actually thinking this could be much better than this pattern <3